task/monorail-server/src/server.rs

```diff
@@ -643,6 +643,37 @@ impl<'a, R: Vsc7448Rw> idl::InOrderMonorailImpl for ServerImpl<'a, R> {
             }
         }
     }
+
+    fn read_vsc7448_mac_count(
+        &mut self,
+        _msg: &userlib::RecvMessage,
+    ) -> Result<usize, RequestError<MonorailError>> {
+        vsc7448::mac::count_macs(self.vsc7448.rw)
+            .map_err(MonorailError::from)
+            .map_err(RequestError::from)
+    }
+
+    fn read_vsc7448_next_mac(
+        &mut self,
+        _msg: &userlib::RecvMessage,
+    ) -> Result<MacTableEntry, RequestError<MonorailError>> {
+        let mac = vsc7448::mac::next_mac(self.vsc7448.rw)
+            .map_err(MonorailError::from)
+            .map_err(RequestError::from)?;
+        // Handle the case where there is no next MAC, e.g. because of a TOCTOU
+        // race with `read_vsc7448_mac_count`
```

Comment:
Two questions:

* Is there a reason we can't return `Result<Option<MacTableEntry>, _>` instead translating a `None` into a sentinel value?
* A TOCTOU issue where the number of mac entries decreased will be visible with this code, but a TOCTOU issue where the number of mac entries increased won't, right? Do we care?

Reply:
- I don't trust `ssmarshal` to properly pack an `Option<MacTableEntry>`, since that's called out in the docs as unallowed.  
- Correct – if the system learns more MAC addresses in between `count_macs` and `next_mac`, they wouldn't be returned.  I think that's fine, since it's morally equivalent to having called the functions _slightly_ sooner (i.e. before the new address was learned).  The one downside is that it would leave the iterator pointing to a non-zero MAC entry, but `count_macs` clears it before running, so that should be okay.

Reply:
* Should we tack a "TODO: switch to Option<_> if/when we fully switch from `ssmarshal` to `hubpack`" somewhere?
* Makes sense :+1:

Reply:
Good idea, done!